daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala

```diff
@@ -416,21 +416,36 @@ class TypingSpec extends WordSpec with TableDrivenPropertyChecks with Matchers {
         E"Λ (τ₁ : ⋆) (τ₂ : ⋆) (τ₃: ⋆). λ (e: ⟨ f₁: τ₁, f₂: τ₂ ⟩) (e₃: τ₃)  → ⸨ ⟨ e with f₂ = e₃ ⟩ ⸩" -> //
           { case _: ETypeMismatch => },
         // ExpCaseVariant
-        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of Mod:Tree:Node x -> () ⸩" -> //
+        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of Mod:Tree:Node x -> () | _ -> () ⸩" -> //
           { case _: EPatternTypeMismatch => },
+        E"Λ (τ : ⋆). λ (e : Mod:Tree τ) → ⸨ case e of Mod:Tree:Node x -> () ⸩" -> //
+          { case _: ENonExhaustivePatterns => },
+        // ExpCaseEnum
+        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of Mod:Color:Red -> () | _ -> () ⸩" -> //
+          { case _: EPatternTypeMismatch => },
+        E"λ (e : Mod:Color) → ⸨ case e of Mod:Color:Red -> () | Mod:Color:Green -> () ⸩" -> //
+          { case _: ENonExhaustivePatterns => },
         // ExpCaseNil
-        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of Nil → () ⸩" -> //
+        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of Nil → () | _ -> () ⸩" -> //
           { case _: EPatternTypeMismatch => },
+        E"Λ (τ : ⋆). λ (e : List τ) → ⸨ case e of Nil → () ⸩" -> //
+          { case _: ENonExhaustivePatterns => },
         // ExpCaseCons
-        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of Cons x y → () ⸩" -> //
+        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of Cons x y → () | _ -> () ⸩" -> //
           { case _: EPatternTypeMismatch => },
-        E"Λ (τ : ⋆). λ (e: List τ) → ⸨ case e of Cons x x → () ⸩" -> //
+        E"Λ (τ : ⋆). λ (e: List τ) → ⸨ case e of Cons x x → () | _ -> () ⸩" -> //
           { case _: EClashingPatternVariables => },
+        E"Λ (τ : ⋆). λ (e : List τ) → ⸨ case e of Cons x y → () ⸩" -> //
+          { case _: ENonExhaustivePatterns => },
         // ExpCaseFalse & ExpCaseTrue
-        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of True → () ⸩" -> //
+        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of True → () | _ -> () ⸩" -> //
           { case _: EPatternTypeMismatch => },
-        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of False → () ⸩" -> //
+        E"Λ (τ : ⋆). λ (e : τ) → ⸨ case e of False → () | _ -> () ⸩" -> //
           { case _: EPatternTypeMismatch => },
+        E"λ (e : Bool) → ⸨ case e of True → () ⸩" -> //
+          { case _: ENonExhaustivePatterns => },
+        E"λ (e : Bool) → ⸨ case e of False → () ⸩" -> //
+          { case _: ENonExhaustivePatterns => },
```

Comment:
We probably need a couple more tests. Not only tests checking that we reject non-exhaustive pattern matches but also tests that make sure we don't reject anything that is exhaustive. I'm a bit afraid of accidentally rejecting DARs where all pattern matches are exhaustive. This would be particularly bad if people upgrade their DAML engine and afterward can't load packages anymore that were loadable before.

Reply:
I will add some tests for that.